### PR TITLE
gh-294: Vendor lib2to3

### PR DESCRIPTION
**Pull request by corona10**
* Use lib2to3 of CPython latest 3.12 version
* Vendor lib2to3 except *.pickle files. (I am not sure it will affect to benchmark result)
* Remove deprecated warning for lib2to3

closes: gh-294

### CODE REVIEW

**Comment by corona10** [comment]
working_solution

File: pyperformance/data-files/benchmarks/bm_2to3/vendor/setup.py
Comment (CAM-Gerlach) [inline pyperformance/data-files/benchmarks/bm_2to3/vendor/setup.py:1]: Instead of using a legacy `setup.py` here, you should be able to replace this with a `pyproject.toml` instead. The following will add the necessary `build-system` config (which you should always have regardless, even if you do use a `setup.py`), and reproduce this same metadata and config.

```toml
[build-system]
requires = ["setuptools>=61"]
build-backend = "setuptools.build_meta"

[project]
name = "lib2to3"
version = "3.12"
description = "lib2to3 for Python 3.13+"

[tool.setuptools.package-data]
"*" = ["*.txt"]
```

I tested this locally on a source build of the latest Python 3.13, with `pyperf`, etc. installed in a clean venv, running `run_benchmark.py` from a checkout of your branch with the `setup.py` deleted and the above `pyproject.toml`, and everything appeared to work just the same, just using modern packaging.

The one other thing you might want to consider is using the recommended `src` directory layout (moving `lib2to3` inside a `src` directory, and adding to the above

```toml
[tool.setuptools.packages.find]
where = ["src"]
```

which makes package discovery more robust and avoids any chance of picking up the local copy when importing, though both should be pretty much non-issues for this specific vendored use case.

**Review by CAM-Gerlach — COMMENTED** [review]
One significant packaging-related recommendation, as well as one other minor comment.

**Comment by corona10** [comment]
@CAM-Gerlach Thanks done!

**Comment by corona10** [comment]
> test / ubuntu-latest - pypy-3.8 (pull_request) 

Not sure which is the root cause.

`````
FAIL: test_run_and_show (test_commands.FullStackTests)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/runner/work/pyperformance/pyperformance/pyperformance/tests/test_commands.py", line 151, in test_run_and_show
    text = self.run_pyperformance(
  File "/home/runner/work/pyperformance/pyperformance/pyperformance/tests/test_commands.py", line 76, in run_pyperformance
    self.assertEqual(ec, exitcode, repr(stdout))
AssertionError: 1 != 0 : None
`````

**Comment by AlexWaygood** [comment]
> > test / ubuntu-latest - pypy-3.8 (pull_request)
> 
> Not sure which is the root cause.
> 
> ```
> FAIL: test_run_and_show (test_commands.FullStackTests)
> ----------------------------------------------------------------------
> Traceback (most recent call last):
>   File "/home/runner/work/pyperformance/pyperformance/pyperformance/tests/test_commands.py", line 151, in test_run_and_show
>     text = self.run_pyperformance(
>   File "/home/runner/work/pyperformance/pyperformance/pyperformance/tests/test_commands.py", line 76, in run_pyperformance
>     self.assertEqual(ec, exitcode, repr(stdout))
> AssertionError: 1 != 0 : None
> ```

The `docutils` benchmark seems to be pretty flaky on PyPy for some reason. I have no idea why, but I've noticed it failing on a lot of PRs and pushes recently (but not on all of them, just some of them).

I'm pretty sure it's unrelated to this PR :)

**Comment by corona10** [comment]
> I'm pretty sure the failure is unrelated to this PR, anyway :)

Thank you Alex for checking this issue :)

File: pyperformance/data-files/benchmarks/bm_2to3/vendor/pyproject.toml
Comment (CAM-Gerlach) [inline pyperformance/data-files/benchmarks/bm_2to3/vendor/pyproject.toml:8]:
```suggestion
description = "lib2to3 for Python 3.13+"
requires-python = ">=3.13.0a0"
license = {text = "PSF-2.0"}
classifiers = [
    "License :: OSI Approved :: Python Software Foundation License",
]
```

Upon further thought, there's a couple bits of additional metadata that you should also include in the `pyproject.toml`, specifically the license stuff (discussed in the top-level comment) and `requires-python`; this will avoid it accidentally being installed on a lower Python version than 3.12, either through a bug, user error, or if this were to ever get distributed elsewhere (by you or someone else) in the future.

**Review by CAM-Gerlach — COMMENTED** [review]
**Standard reminder: You can directly apply all the suggestions you want in one go by going to [`Files changed`](URL) -> Clicking <kbd>Add to batch</kbd> on each suggestion -> When done, clicking <kbd>Commit</kbd>**

One thing I forgot to mention—to legally include the `lib2to3` code here, you'll need to include the PSF v2 license under which `lib2to3` is licensed (as a module added in Python >2.2+ with no documented additional terms) to a `LICENSE.txt` file in the `vendor` directory (which will get picked up by Setuptools and Wheel automatically), and also add the appropriate license identifier and classifiers to the metadata (which I added review suggestions for). Here is the exact relevant license reproduced from the CPython `LICENSE` file:

<details>

<summary>License text you need to add to a `LICENSE.txt` file in `vendor`</summary>

```text
PYTHON SOFTWARE FOUNDATION LICENSE VERSION 2
--------------------------------------------

1. This LICENSE AGREEMENT is between the Python Software Foundation
("PSF"), and the Individual or Organization ("Licensee") accessing and
otherwise using this software ("Python") in source or binary form and
its associated documentation.

2. Subject to the terms and conditions of this License Agreement, PSF hereby
grants Licensee a nonexclusive, royalty-free, world-wide license to reproduce,
analyze, test, perform and/or display publicly, prepare derivative works,
distribute, and otherwise use Python alone or in any derivative version,
provided, however, that PSF's License Agreement and PSF's notice of copyright,
i.e., "Copyright (c) 2001, 2002, 2003, 2004, 2005, 2006, 2007, 2008, 2009, 2010,
2011, 2012, 2013, 2014, 2015, 2016, 2017, 2018, 2019, 2020, 2021, 2022, 2023 Python Software Foundation;
All Rights Reserved" are retained in Python alone or in any derivative version
prepared by Licensee.

3. In the event Licensee prepares a derivative work that is based on
or incorporates Python or any part thereof, and wants to make
the derivative work available to others as provided herein, then
Licensee hereby agrees to include in any such work a brief summary of
the changes made to Python.

4. PSF is making Python available to Licensee on an "AS IS"
basis.  PSF MAKES NO REPRESENTATIONS OR WARRANTIES, EXPRESS OR
IMPLIED.  BY WAY OF EXAMPLE, BUT NOT LIMITATION, PSF MAKES NO AND
DISCLAIMS ANY REPRESENTATION OR WARRANTY OF MERCHANTABILITY OR FITNESS
FOR ANY PARTICULAR PURPOSE OR THAT THE USE OF PYTHON WILL NOT
INFRINGE ANY THIRD PARTY RIGHTS.

5. PSF SHALL NOT BE LIABLE TO LICENSEE OR ANY OTHER USERS OF PYTHON
FOR ANY INCIDENTAL, SPECIAL, OR CONSEQUENTIAL DAMAGES OR LOSS AS
A RESULT OF MODIFYING, DISTRIBUTING, OR OTHERWISE USING PYTHON,
OR ANY DERIVATIVE THEREOF, EVEN IF ADVISED OF THE POSSIBILITY THEREOF.

6. This License Agreement will automatically terminate upon a material
breach of its terms and conditions.

7. Nothing in this License Agreement shall be deemed to create any
relationship of agency, partnership, or joint venture between PSF and
Licensee.  This License Agreement does not grant permission to use PSF
trademarks or trade name in a trademark sense to endorse or promote
products or services of Licensee, or any third party.

8. By copying, installing or otherwise using Python, Licensee
agrees to be bound by the terms and conditions of this License
Agreement.
```

</details>

**Comment by CAM-Gerlach** [comment]
@corona10 If you want to install it from the parent `pyproject.toml` as you wanted to do rather than part of the script, that's definitely be best as you say.

It's not really easily possible to do it truly locally, at least in a standards-conformant, non-hacky/bespoke way. You could add that subdirectory to the package finder to install it as another top-level package, but there's no way to make that conditional on a specific Python version without using a custom, backend-specific dynamic `setup.py`.  Adding it as a dependency with the above marker allows it to be conditional on Python version and installed separately, and makes more sense, but to do that with a local path requires a non-standard, backend-bespoke mechanism (e.g. like [this](https://github.com/timkittel/stackoverflow_demo/blob/main/pyproject.toml 
) with hatchling, or possibly with `setup.py` hacks with Setuptools) and is apparently not compatible with `pip` and possibly other frontends (at least not yet).

However, there _are_ several options to still approximate this, and what seems the least work and lowest friction is to specify it as a dependency using a [VCS URL reference](https://pip.pypa.io/en/stable/topics/vcs-support/#url-fragments) to this online repository, with the appropriate subdirectory, and git tag specified (which to test the PR prior to merging, will need to be the appropriate branch on your fork), and with a PEP 508 marker limiting it to Python 3.13+. I.e., the `dependencies` in `bm_2to3/pyproject.toml` would be:

```toml
dependencies = ["pyperf", 'lib2to3 @ https://github.com/python/pyperformance.git@2to3v3.12#subdirectory=pyperformance/data-files/benchmarks/bm_2to3/vendor ; python_version >= "3.13.0a0"']
```

This requires minimal extra work up front, though it does make development a bit more tedious if you ever want to make changes to `lib2to3` (e.g. for compatibility with a new Python runtime version), as you'll need to push your changes to your fork, modify the repo and ref to test them, and then remember to point it back to a new git tag on the upstream before merging (if you don't specify a tag, existing versions won't be reproducible and will break if you e.g. change any part of the path, or make any incompatible changes to `lib2to3` itself). I initially was going to suggest just using the tag of the pyperformance release, but it makes a lot more sense to just use a separate tag for `lib2to3` as its easier and more flexible overall.

Alternatively, you could host it in its own separate repo, but I'm not sure that would really offer much benefit over just using a separate tag for `lib2to3` changes, compared to the additional overhead and other costs. And/or, you could upload it as a forward-port to PyP which would simplify the spec, but otherwise add even more complexity for not much benefit (other than a little more long-term stability). So, I'd recommend the above approach assuming you want to avoid the `run_benchmark,py` hacks here.

**Comment by CAM-Gerlach** [comment]
Sidenote: You're missing the required `[build-system]` section in that `pyproject.toml`—might want to add that as appropriate (unless there's a specific reason it's missing?). E.g. if you're using Setuptools:

```toml
[build-system]
requires = ["setuptools>=61"]
build-backend = "setuptools.build_meta"
```

**Review by CAM-Gerlach — APPROVED** [review]
LGTM from me as far as a packaging is concerned :+1:

**Comment by corona10** [comment]
Waiting for @gvanrossum 's final review as the 2to3 author and vendoring suggester.

File: pyperformance/data-files/benchmarks/bm_2to3/run_benchmark.py
Comment (gvanrossum) [reply pyperformance/data-files/benchmarks/bm_2to3/run_benchmark.py:1]: What he said.

**Review by gvanrossum — COMMENTED** [review]
I don’t have much expertise here, if CAM is okay you can merge it.

File: pyperformance/data-files/benchmarks/bm_2to3/run_benchmark.py
Comment (corona10) [reply pyperformance/data-files/benchmarks/bm_2to3/run_benchmark.py:1]: > Since you mentioned wanting to avoid this hack and wanting to install it with pyproject.toml instead, I thought some more about this and left a comment below explaining how you should be able to avoid all the changes to this file and instead just add lib2to3 it as a dependency to the bm_2to3/pyproject.toml.

> What he said.

As I commented on the discord, VCS approach is the most approximated way to handle this, but I will keep the current approach since managing the VCS URL has some issues when we want to modify the codes and the bootstrap(?) problem.

**Pull request merged**
